### ISSUE TITLE
"Access to protected property" incorrect warnings [NTI]

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#1943]
I'm getting 383 warnings in my large codebase, probably since @aravind-pg's https://github.com/google/closure-compiler/commit/1ab80b484fb666ec681e3edb99096adeb4b84787.

I have a couple of scripts (i.e. goog provided, not a module) which declare classes using the traditional `@constructor` annotated function and prototype properties. One class declares an `@protected' member. Another directly extends it.

The warnings appear **bizarrely not on the first access to the property, but a few accesses later**, e.g. 2nd or 3rd and 4th accesses, per enclosing file.

**Comment by supersteves** [comment]
[Test case in the debugger!](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.provide(%2522A%2522)%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250AA%2520%253D%2520function()%2520%257B%250A%250A%2520%2520%2520%2520%252F**%250A%2520%2520%2520%2520%2520*%2520%2540type%2520%257BElement%257D%250A%2520%2520%2520%2520%2520*%2520%2540protected%250A%2520%2520%2520%2520%2520*%252F%250A%2520%2520%2520%2520this._element%2520%253D%2520null%253B%250A%257D%250A%250A%26input1%3Dgoog.provide(%2522B%2522)%253B%250A%250Agoog.require(%2522A%2522)%253B%250A%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%2520%2540extends%2520%257BA%257D%250A%2520*%252F%250AB%2520%253D%2520function()%2520%257B%250A%2520%2520%2520%2520A.call(this)%253B%250A%2520%2520%2520%2520this._element.appendChild(document.createElement(%2522div%2522))%253B%250A%250A%2520%2520%2520%2520this._element.appendChild(document.createElement(%2522span%2522))%253B%250A%257D%253B%250A%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CHECK_VISIBILITY%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1)

**Comment by supersteves** [comment]
Only happens with NTI turned on. With OTI, there is no error, but if I change it to `@private` the access check correctly generates an error, so it's clearly still running the check. It may be something in NTI rather than in CheckAccessControls.

**Comment by aravind-pg** [comment]
Thanks for the bug report!  I have a fix for this that should land by tomorrow.

FWIW, what was happening was that we were using the *first* property access on `this._element` to tighten its inferred type from `Object` to `!Object`, but in doing so lost the information that `_element` was an inherited property (due to an implementation detail). This then manifested itself only in the *second* and later property accesses :)

**Comment by supersteves** [comment]
Thanks guys. I'll confirm the fix once we get past https://github.com/google/closure-compiler/issues/1958.

**Comment by supersteves** [comment]
Confirmed. Thanks!

**Comment by supersteves** [comment]
Nope. I spoke too soon. Was running without NTI. With NTI, my codebase now generates 56 such warnings - a vast improvement on 380 warnings, but I've confirmed these are still incorrect warnings. Please re-open.

I'm unable to check the debugger link (above) to see if that case is fixed, because it's timing out after a couple of minutes. Is there any way this debugger can be improved to be more responsive? Even when it doesn't timeout it's very tedious trying to reproduce a test case with it.

I glanced through the errors, and spot checked a couple in detail. It seems most, perhaps all, relate to a common superclass way up in the hierarchy, about 5 superclasses up. Both checked cases referred to protected members declared in classes which directly extend this common superclass. The warnings were raised about 3 subclasses down. These are all goog.provide-based scripts rather than ES6 or goog. modules.

**Comment by dimvar** [comment]
If you manage to provide us with a repro, that will help a lot.

**Comment by supersteves** [comment]
It's my last day before a 2 week break, so probably not right now. 

Meanwhile is there any chance the [debugger](https://closure-compiler-debugger.appspot.com/) can be given more resource, because it's near unusable, and I use this all the time when trying to narrow down a compiler bug?

**Comment by dimvar** [comment]
Regarding repro, we ran into a similar warning yesterday in internal code, so I'm guessing the fix for that will fix the issue you're seeing as well.

**Comment by MatrixFrog** [comment]
What's unusuable about it? It's too slow? There might be some way to tell appengine to give it more CPU/memory/something, I don't know. There is also https://closure-compiler-debugger.appspot.com/gwt_debugger.html which tends to be faster, but doesn't (yet) include the default externs so you often have to type in externs (at least for things like Array and Function) yourself.

**Comment by supersteves** [comment]
I didn't know about gwt_debugger, I'll play with that next time.

Yes, I imagine someone can probably pull some lever to make it go 10x faster... I hope!

So basically it's always been slow to respond, often taking say 30 seconds to compile and show the output. But in the last 6 months or thereabouts, it's been consistently slower and today I tried it several times and every time it timed out with an exception.  If you're playing with compiler options or syntax, you need it to be responsive, otherwise it kills your productivity/focus, since you need to make tiny changes and check results, many times over. 

I much prefer creating a test case online in the debugger, than doing one locally, because I don't have to set up the build script/command and source files (we have our own highly convoluted build process), and I want to be able to post a link you can just click to (ideally) instantly see if it's still broken / to demonstrate to the compiler authors. One thing (other than performance) it's also missing is a way of creating test cases with goog.module or ES6 modules.

**Comment by aravind-pg** [comment]
@supersteves I've just spent a bunch of time looking into some similar issues internally, and while I can't be sure the issues you're facing are the same, here are some thoughts.

The main thing is: are you seeing "Invalid property redeclaration" NTI warnings in your codebase? If so that's a red flag. I have a fairly strong suspicion that's the root cause for these CheckAccessControls warnings. So I'd be very interested in knowing if fixing those warnings causes the CheckAccessControls warnings to go away.

But there's more to this. In some cases the "Invalid property redeclaration" warnings are caused by coding convention issues. Specifically, NTI implicitly relies on some behavior of GoogleCodingConvention (such as treating opt_ parameters as optional) which e.g. ClosureCodingConvention does not have. Can you check which convention you are using in your codebase? If it is not GoogleCodingConvention then that'd explain some things.

If it does turn out to be purely a coding convention thing, then I'm not yet sure what we'll do with that situation. Ideally we'd like to avoid reliance on any particularly coding convention, but it's TBD how exactly to go about that.

**Comment by supersteves** [comment]
I don't think we were getting that other error but can't be sure. Definitely using Closure coding convention not Google so that may explain it. Will look further when back.

**Comment by aravind-pg** [comment]
@supersteves OK I think I'm fairly sure I found the core issue you're most likely facing: see #1976. It is a legitimate bug with nothing to do with coding conventions.

**Comment by dimvar** [comment]
Hi Steve, I've started looking into this issue. Can you take a look at Aravind's first example in #1976 and see if it has the same shape as your code that causes the warning? If not, can you provide a repro for the remaining warnings in your code?

**Comment by supersteves** [comment]
@dimvar Can't recall exactly but I'm pretty sure this is a direct inheritance issue purely to do with looking up through the superclass chain when calling this.something.  So I don't think it's Aravind's case of anotherObject.something.  Happy to run any test/diagnostic code you give me or do a spot of debugging if directed?
